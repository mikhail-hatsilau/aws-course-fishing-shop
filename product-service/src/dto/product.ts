import { array, boolean, InferType, number, object, string } from 'yup';

export const productImageSchema = object({
  id: string().uuid().required().default(''),
  main: boolean(),
  source: string().required().default(''),
});

export const createProductImageRequestSchema = object({
  main: boolean(),
  source: string().required().default(''),
});

export type ProductImage = InferType<typeof productImageSchema>;

export const productSchema = object({
  id: string().uuid().required().default(''),
  title: string().required().default(''),
  description: string().required().default(''),
  price: number().required(),
  categoryId: string().uuid().required().default(''),
  count: number().required(),
  images: array(productImageSchema).required().default([]),
});

export type Product = InferType<typeof productSchema>;

export const createProductRequestSchema = object({
  title: string().required().default(''),
  description: string().required().default(''),
  price: number().required(),
  categoryId: string().uuid().required().default(''),
  count: number().required(),
  images: array(createProductImageRequestSchema).required().default([]),
});

export type CreateProductRequest = InferType<typeof createProductRequestSchema>;

export interface ProductRow {
  id: string;
  title: string;
  description: string;
  price: number;
  category_id: string;
  count: number;
  image_id: string;
  image_source: string;
  main_image: boolean;
}
