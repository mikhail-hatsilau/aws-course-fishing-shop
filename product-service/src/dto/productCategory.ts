export interface ProductCategory {
  id: string;
  title: string;
  subCategoryId: string;
}
