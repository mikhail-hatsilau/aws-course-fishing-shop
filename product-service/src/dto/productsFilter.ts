import { InferType, object, string } from 'yup';

export const getProductsSchema = object({
  categoryId: string().uuid(),
});

export const getProductByIdPathParamsSchema = object({
  id: string().uuid().required(),
});

export type GetProductByIdPathParams = InferType<
  typeof getProductByIdPathParamsSchema
>;

export type GetProductsFilter = InferType<typeof getProductsSchema>;
