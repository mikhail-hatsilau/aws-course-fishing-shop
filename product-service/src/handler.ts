export { getProductsList } from './handlers/getProductsList.handler';
export { getProductById } from './handlers/getProductById.handler';
export { createProduct } from './handlers/createProduct.handler';
export { catalogBatchProcess } from './handlers/catalogBatchProcess.handler';
