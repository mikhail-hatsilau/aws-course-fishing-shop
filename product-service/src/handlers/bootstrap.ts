import { NestFactory } from '@nestjs/core';
import { AppModule } from '../app.module';

export const bootstrap = () => {
  return NestFactory.createApplicationContext(AppModule, {
    logger: ['error'],
  });
};
