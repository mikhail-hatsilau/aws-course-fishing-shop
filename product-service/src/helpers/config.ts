import { Injectable } from '@nestjs/common';
import { EnvVariables } from 'src/types/envVariables';

@Injectable()
export class Config {
  getEnvVariable(name: keyof EnvVariables) {
    return process.env[name] as EnvVariables[typeof name];
  }
}
