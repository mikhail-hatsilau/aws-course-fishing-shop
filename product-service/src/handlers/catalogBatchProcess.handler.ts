import middy from '@middy/core';
import inputOutputLogger from '@middy/input-output-logger';
import { SQSEvent } from 'aws-lambda';
import { bootstrap } from './bootstrap';
import { YupValidationService } from '../services/yupValidation.service';
import { ValidationService } from '../services/validation.abstract.service';
import { createProductRequestSchema, Product } from '../dto/product';
import { ProductsService } from '../services/products.interface.service';
import { DefaultProductsService } from '../services/products.service';
import { ProductError } from '../helpers/errors/productError';
import { ProductNotificationService } from '../services/productNotification.interface.service';
import { DefaultProductNotificationService } from '../services/productNotification.service';

const catalogBatchProcessHandler = async (event: SQSEvent) => {
  const app = await bootstrap();
  const validationService = app.get<YupValidationService>(ValidationService);
  const productsService = app.get<ProductsService>(DefaultProductsService);
  const productNotificationService = app.get<ProductNotificationService>(
    DefaultProductNotificationService,
  );
  const { Records: productRecords } = event;
  const products = productRecords.map((record) => JSON.parse(record.body));
  const validProducts: Product[] = [];

  console.log('Validate products...');
  try {
    await Promise.all(
      products.map(async (product) => {
        await validationService.validate(createProductRequestSchema, product);
        validProducts.push(product);
      }),
    );
  } catch (error) {
    console.error(error);
  }
  console.log(`Identified ${validProducts.length} of valid products`);
  console.log('Creating products...');

  const addedProducts: Product[] = [];
  await Promise.all(
    validProducts.map(async (product) => {
      try {
        addedProducts.push(await productsService.createProduct(product));
      } catch (e) {
        if (e instanceof ProductError) {
          console.error(e);
          return;
        }
        throw e;
      }
    }),
  );

  if (addedProducts.length) {
    console.log('Products successfully created!');
    console.log('Sending notification...');
    await productNotificationService.notifyProductsCreated(addedProducts);
    console.log('Notification sent successfully!');
  }
};

export const catalogBatchProcess = middy(catalogBatchProcessHandler).use(
  inputOutputLogger(),
);
