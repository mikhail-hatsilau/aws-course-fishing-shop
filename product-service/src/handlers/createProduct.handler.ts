import { HttpStatus } from '@nestjs/common';
import middy from '@middy/core';
import cors from '@middy/http-cors';
import httpJsonBodyParser, { Event } from '@middy/http-json-body-parser';
import { HTTPResponse } from '../helpers/response/httpResponse';
import { bootstrap } from './bootstrap';
import httpResponseSerializer from '@middy/http-response-serializer';
import inputOutputLogger from '@middy/input-output-logger';
import { defaultSerializers } from '../helpers/serializers';
import { ValidationService } from '../services/validation.abstract.service';
import { YupValidationService } from '../services/yupValidation.service';
import {
  CreateProductRequest,
  createProductRequestSchema,
} from '../dto/product';
import { ProductsService } from '../services/products.interface.service';
import { DefaultProductsService } from '../services/products.service';

const createProductHandler = async (event: Event) => {
  try {
    const { body } = event;
    const app = await bootstrap();
    const validationService = app.get<YupValidationService>(ValidationService);

    try {
      await validationService.validate(createProductRequestSchema, body);
    } catch (error) {
      return new HTTPResponse(HttpStatus.BAD_REQUEST, error);
    }

    const product = body as CreateProductRequest;
    const productsService = app.get<ProductsService>(DefaultProductsService);
    const createdProduct = await productsService.createProduct(product);

    return new HTTPResponse(HttpStatus.OK, createdProduct);
  } catch (e) {
    console.error('Error encountered: ', e);
    return new HTTPResponse(
      HttpStatus.INTERNAL_SERVER_ERROR,
      'Internal Server Error',
    );
  }
};

export const createProduct = middy(createProductHandler)
  .use(cors())
  .use(inputOutputLogger())
  .use(httpJsonBodyParser())
  .use(
    httpResponseSerializer({
      serializers: defaultSerializers,
      defaultContentType: 'application/json',
    }),
  );
