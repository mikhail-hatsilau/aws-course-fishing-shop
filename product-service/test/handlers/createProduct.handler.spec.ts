import { Product } from '../../src/dto/product';
import { bootstrap } from '../../src/handlers/bootstrap';
import { createProduct } from '../../src/handlers/createProduct.handler';
import { DefaultProductsService } from '../../src/services/products.service';
import { ProductsRepository } from '../../src/repositories/products.abstract.repository';
import { MockProductsRepository } from '../mocks/productsMock.repository';
import { Config } from '../../src/helpers/config';
import { ValidationService } from '../../src/services/validation.abstract.service';
import { YupValidationService } from '../../src/services/yupValidation.service';
import { Test } from '@nestjs/testing';
import { Event } from '@middy/http-json-body-parser';

jest.mock('../../src/handlers/bootstrap', () => {
  return {
    bootstrap: jest.fn(),
  };
});

const mockEvent: Event = {
  rawBody: '{}',
  body: {},
  resource: 'products/',
  path: 'products/',
  httpMethod: 'POST',
  isBase64Encoded: false,
  queryStringParameters: {},
  multiValueQueryStringParameters: {},
  pathParameters: {},
  stageVariables: {},
  headers: {},
  multiValueHeaders: {},
  requestContext: {
    authorizer: null,
    accountId: '123456789012',
    resourceId: '123456',
    stage: 'dev',
    requestId: 'c6af9ac6-7b61-11e6-9a41-93e8deadbeef',
    requestTime: '09/Apr/2015:12:34:56 +0000',
    requestTimeEpoch: 1428582896000,
    identity: {
      cognitoIdentityPoolId: null,
      accountId: null,
      cognitoIdentityId: null,
      caller: null,
      accessKey: null,
      sourceIp: '127.0.0.1',
      apiKey: '',
      apiKeyId: '',
      clientCert: null,
      principalOrgId: '',
      cognitoAuthenticationType: null,
      cognitoAuthenticationProvider: null,
      userArn: null,
      userAgent: 'Custom User Agent String',
      user: null,
    },
    path: 'products/',
    resourcePath: 'products/',
    httpMethod: 'GET',
    apiId: '1234567890',
    protocol: 'HTTP/1.1',
  },
};

const context = {
  callbackWaitsForEmptyEventLoop: true,
  functionName: 'createProduct',
  functionVersion: '1',
  invokedFunctionArn: 'arn',
  memoryLimitInMB: '100',
  awsRequestId: 'c6af9ac6-7b61-11e6-9a41-93e8deadbeef',
  logGroupName: 'logGroup',
  logStreamName: 'streamName',
  getRemainingTimeInMillis: jest.fn(),
  done: jest.fn(),
  fail: jest.fn(),
  succeed: jest.fn(),
};

const productMock = {
  description: 'Salmo Extreme BP Feeder 090 3.60 / 3133-360',
  id: '7567ec4b-b10c-48c5-9345-fc73c48a80aa',
  price: 110,
  title: 'Salmo Extreme BP Feeder 090 3.60 / 3133-360',
  categoryId: 'c8c1770a-abb1-4dba-92e0-07528757db58',
  count: 1,
  images: [],
};

describe('createProductHandler', () => {
  let createProductSpy: jest.SpyInstance<Promise<Product | undefined>>;

  beforeEach(async () => {
    (bootstrap as jest.Mock).mockImplementation(async () => {
      const moduleRef = await Test.createTestingModule({
        controllers: [],
        providers: [
          Config,
          DefaultProductsService,
          { provide: ValidationService, useClass: YupValidationService },
          { provide: ProductsRepository, useClass: MockProductsRepository },
        ],
      }).compile();
      return moduleRef;
    });
    createProductSpy = jest.spyOn(
      DefaultProductsService.prototype,
      'createProduct',
    );
  });

  afterEach(() => {
    createProductSpy.mockRestore();
  });

  it('should create product successfully', async () => {
    createProductSpy.mockImplementation(() => Promise.resolve(productMock));
    const event = {
      ...mockEvent,
      rawBody: JSON.stringify(productMock),
      body: productMock,
    };
    const response = await createProduct(event, context);

    expect(response).toEqual(
      expect.objectContaining({
        statusCode: 200,
        body: JSON.stringify(productMock),
      }),
    );
  });

  it('should set CORS headers', async () => {
    createProductSpy.mockImplementation(() => Promise.resolve(productMock));
    const response = await createProduct(mockEvent, context);

    expect(response).toEqual(
      expect.objectContaining({
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        },
      }),
    );
  });

  it('should return 400 when incorrect product is passed', async () => {
    createProductSpy.mockImplementation(() => Promise.resolve(undefined));

    const incorrectProduct = {
      description: 'Salmo Extreme BP Feeder 090 3.60 / 3133-360',
      categoryId: 'c8c1770a-abb1-4dba-92e0-07528757db58',
      count: 1,
      images: [],
    };
    const response = await createProduct(
      {
        ...mockEvent,
        rawBody: JSON.stringify(incorrectProduct),
        body: incorrectProduct,
      },
      context,
    );

    expect(response).toMatchSnapshot();
  });
});
