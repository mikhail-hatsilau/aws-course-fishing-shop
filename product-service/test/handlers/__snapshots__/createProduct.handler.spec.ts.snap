// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProductHandler should return 400 when incorrect product is passed 1`] = `
HTTPResponse {
  "body": "{\\"value\\":{\\"images\\":[],\\"count\\":1,\\"categoryId\\":\\"c8c1770a-abb1-4dba-92e0-07528757db58\\",\\"description\\":\\"Salmo Extreme BP Feeder 090 3.60 / 3133-360\\",\\"title\\":\\"\\"},\\"path\\":\\"price\\",\\"type\\":\\"required\\",\\"errors\\":[\\"price is a required field\\"],\\"params\\":{\\"path\\":\\"price\\"},\\"inner\\":[],\\"name\\":\\"ValidationError\\",\\"message\\":\\"price is a required field\\"}",
  "cookies": undefined,
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "isBase64Encoded": undefined,
  "statusCode": 400,
}
`;
